toms462.c

```diff
@@ -0,0 +1,634 @@
+     3.0, -0.2, -0.2, -0.2, -0.2,
+    -0.2, -0.2, -0.2, -0.2, -0.2,
+    -0.2, -0.2, -0.2, -0.2, -0.2,
+    -0.2, -0.2, -0.2, -0.2, -0.2,
+     1.0,  2.0,  0.0,  0.0,  0.0,
+     0.0,  0.0,  1.0,  1.0, -1.0,
+    -1.0,  1.0,  1.0, -1.0, -1.0,
+     0.7071067811865475 };
+  static double y_vec[N_MAX] = {
+     1.0,  1.0,  1.0,  1.0,  1.0,
+     1.0,  0.5,  0.5,  0.5,  0.5,
+     0.5,  0.5,  0.5,  0.5,  0.5,
+     0.5,  0.5,  0.5,  0.5,  0.5,
+     0.5,  0.5,  0.5,  0.5,  0.5,
+     0.5,  1.0,  0.0,  0.0,  0.0,
+     0.0,  0.0,  1.0, -1.0,  1.0,
+    -1.0,  1.0, -1.0,  1.0, -1.0,
+     0.7071067811865475 };
+
+  if ( *n_data < 0 )
+  {
+    *n_data = 0;
+  }
+
+  *n_data = *n_data + 1;
+
+  if ( N_MAX < *n_data )
+  {
+    *n_data = 0;
+    *r = 0.0;
+    *x = 0.0;
+    *y = 0.0;
+    *fxy = 0.0;
+  }
+  else
+  {
+    *r = r_vec[*n_data-1];
+    *x = x_vec[*n_data-1];
+    *y = y_vec[*n_data-1];
+    *fxy = fxy_vec[*n_data-1];
+  }
+
+  return;
+# undef N_MAX
+}
+/******************************************************************************/
+
+double bivnor ( double ah, double ak, double r )
+
+/******************************************************************************/
+/*
+  Purpose:
+
+    BIVNOR computes the bivariate normal CDF.
+
+  Discussion:
+
+    BIVNOR computes the probability for two normal variates X and Y
+    whose correlation is R, that AH <= X and AK <= Y.
+
+  Licensing:
+
+    This code is distributed under the GNU LGPL license.
+
+  Modified:
+
+    13 April 2012
+
+  Author:
+
+    Original FORTRAN77 version by Thomas Donnelly.
+    C version by John Burkardt.
+
+  Reference:
+
+    Thomas Donnelly,
+    Algorithm 462: Bivariate Normal Distribution,
+    Communications of the ACM,
+    October 1973, Volume 16, Number 10, page 638.
+
+  Parameters:
+
+    Input, double AH, AK, the lower limits of integration.
+
+    Input, double R, the correlation between X and Y.
+
+    Output, double BIVNOR, the bivariate normal CDF.
+
+  Local Parameters:
+
+    Local, int IDIG, the number of significant digits
+    to the right of the decimal point desired in the answer.
+*/
+{
+  double a2;
+  double ap;
+  double b;
+  double cn;
+  double con;
+  double conex;
+  double ex;
+  double g2;
+  double gh;
+  double gk;
+  double gw;
+  double h2;
+  double h4;
+  int i;
+  static int idig = 15;
+  int is;
+  double rr;
+  double s1;
+  double s2;
+  double sgn;
+  double sn;
+  double sp;
+  double sqr;
+  double t;
+  static double twopi = 6.283185307179587;
+  double w2;
+  double wh;
+  double wk;
+
+  b = 0.0;
+
+  gh = gauss ( - ah ) / 2.0;
+  gk = gauss ( - ak ) / 2.0;
+
+  if ( r == 0.0 )
+  {
+    b = 4.00 * gh * gk;
+    b = r8_max ( b, 0.0 );
+    b = r8_min ( b, 1.0 );
+    return b;
+  }
+
+  rr = ( 1.0 + r ) * ( 1.0 - r );
+
+  if ( rr < 0.0 )
+  {
+    fprintf ( stderr, "\n" );
+    fprintf ( stderr, "BIVNOR - Fatal error!\n" );
+    fprintf ( stderr, "  1 < |R|.\n" );
+    exit ( 0 );
```

Comment:
This is absolutely unacceptable in XS-code.

Please, check the pre-conditions in XS code wrapper, to be sure, that they'll never happen (rr < 0), and if they do, then just do not call that function, but die with error.

When you do preconditions check, please, comment (but not remove) this lines, that they are checked.